pkg/winperfcounters/watcher.go

```diff
@@ -0,0 +1,147 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build windows
+// +build windows
+
+package winperfcounters // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/winperfcounters"
+
+import (
+	"fmt"
+
+	"go.uber.org/multierr"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/winperfcounters/internal/pdh"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/winperfcounters/internal/third_party/telegraf/win_perf_counters"
+)
+
+var _ PerfCounterWatcher = (*Watcher)(nil)
+
+// PerfCounterWatcher represents how to scrape data
+type PerfCounterWatcher interface {
+	// Path returns the counter path
+	Path() string
+	// ScrapeData collects a measurement and returns the value(s).
+	ScrapeData() ([]win_perf_counters.CounterValue, error)
+	// Close all counters/handles related to the query and free all associated memory.
+	Close() error
+
+	GetMetricRep() MetricRep
+}
+
+const instanceLabelName = "instance"
+
+type Watcher struct {
+	Counter *pdh.PerfCounter
+	MetricRep
+}
+
+func (w Watcher) Path() string {
+	return w.Counter.Path()
+}
+
+func (w Watcher) ScrapeData() ([]win_perf_counters.CounterValue, error) {
+	return w.Counter.ScrapeData()
+}
+
+func (w Watcher) Close() error {
+	return w.Counter.Close()
+}
+
+func (w Watcher) GetMetricRep() MetricRep {
+	return w.MetricRep
+}
+
+// BuildPaths creates watchers and their paths from configs.
+func BuildPaths(objs []ObjectConfig) ([]PerfCounterWatcher, error) {
```

Comment:
Am I understanding correctly that the input/output of this function is _not_ 1:1, but that the input/output of `WatchCounters` _is_ 1:1?

If so, the only benefit to batching all this into loops is to help the user avoid a fairly simple outer loop. I think this batching complicates the API and the implementation.

IMO, I should interact with this library like this:

```go
func (r myReciever) Start() {
	watchers := []PerfCounterWatcher{}
	for _, objCfg := range r.objCfgs {
		objWatchers, err := objCfg.Build()
		if err != nil { ... }
		for _, objWatcher := range objWatchers {
			watchers = append(watchers, objWatcher)
		}
	}
	r.watchers = watchers
}

func (r myReciever) Scrape() {
	for _, watcher := range r.watchers {
		counterValue, err := watcher.Scrape()
		if err != nil { ... }
		r.mb.RecordAppropriateDataPoint(counterValue...)
	}
}

func (r myReciever) Stop() {
	for _, watcher := range r.watchers {
		err := watcher.Close()
		if err != nil { ... }
	}
}
```

Am I wrong in thinking that basically all this requires is peeling off the outer loop of each of the three main functions in this library?

Reply:
Unwraped the functions